supabase/functions/submit_contact_request/index.ts

```diff
@@ -0,0 +1,189 @@
+
+const SUPABASE_URL = Deno.env.get("URL");
+const SUPABASE_ANON_KEY = Deno.env.get("ANON_KEY");
+const SUPABASE_SERVICE_ROLE_KEY = Deno.env.get("SERVICE_ROLE_KEY");
+
+const handler = async (_request: Request): Promise<Response> => {
+	if (_request.method === "OPTIONS") {
+		return new Response(null, { headers: corsHeaders, status: 204 });
+	}
+
+	const { userContactName, message } = await _request.json();
+
+	const authHeader = _request.headers.get("Authorization")!;
+
+	const supabaseClient = createClient(SUPABASE_URL, SUPABASE_ANON_KEY, {
+		global: { headers: { Authorization: authHeader } },
+	});
+
+	const supabaseServiceRoleClient = createClient(
+		SUPABASE_URL,
+		SUPABASE_SERVICE_ROLE_KEY
+	);
+
+	// Get the user from the token
+	const token = authHeader.replace("Bearer ", "");
+	const { data: authData, error: authError } =
+		await supabaseClient.auth.getUser(token);
+
+	if (authError) {
+		return new Response(undefined, { status: 401 });
+	}
+
+	// Lookup the contact user id
+	const { data: contactData, error: contactError } =
+		await supabaseServiceRoleClient
+			.from("profiles")
+			.select("*")
+			.eq("username", userContactName)
+			.single();
+
+	if (contactError) {
+		return new Response(undefined, { status: 404, headers: corsHeaders });
+	}
+
+	// Check if the user has already tried to contact the contact
+	const { data: lookupData, error: lookupError } = await supabaseClient
+		.from("contact_requests")
+		.select("*")
+		.eq("user_id", authData.user.id)
+		.eq("contact_id", contactData.id)
+		.not("contact_mail_id", "is", null); // only count sent emails
+
+	if (lookupError) {
+		return new Response(undefined, { status: 500, headers: corsHeaders });
+	}
+
+	if (lookupData.length > 0) {
+		return new Response(
+			JSON.stringify({
+				code: "already_sent_contact_request",
+				message:
+					"User has already send a contact request, not allowed to send another one.",
+			}),
+			{
+				status: 200,
+				headers: {
+					...corsHeaders,
+					"Content-Type": "application/json",
+				},
+			}
+		);
+	}
+
+	// Check if the user has sent 3 contact requests in the last 24 hours
+	const { data: lookupDailyData, error: lookupDailyError } =
+		await supabaseClient
+			.from("contact_requests")
+			.select("*")
+			.eq("user_id", authData.user.id)
+			.not("contact_mail_id", "is", null) // only count sent emails
+			.gt("created_at", sub(new Date(), { days: 1 }).toISOString());
+
+	if (lookupDailyError) {
+		console.log(lookupDailyError);
+		return new Response(undefined, { status: 500, headers: corsHeaders });
+	}
+
+	if (lookupDailyData.length >= 3) {
+		return new Response(
+			JSON.stringify({
+				code: "already_sent_more_than_3_contact_requests",
+				message:
+					"User has already sent more than 3 contact requests in the last 24 hours, not allowed to send another one.",
+			}),
+			{
+				status: 200,
+				headers: {
+					...corsHeaders,
+					"Content-Type": "application/json",
+				},
+			}
+		);
+	}
```

Comment:
Not sure I understand. We first check if the sender has already contacted the receiver, if yes, we return an error. But then we check if the sender has sent already three times in the last 24h? Am I getting this right?

Reply:
we want to prevent two scenarios:

1. we block the contact request if the logged in user already contacted the requested user
2. we block all contact requests if the user already contacted >=3 distinct users in the last 24h

Reply:
Okay I think I got it now. I think the terms user and contact are/were confusing me a bit. For more clarity I'd suggest to use a different naming convention in this file, e.g. sender/receiver, or something else if you have a better idea 👍

Reply:
tried to find better variable names in last commit👍🏼